File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/08_StringtoInteger.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    MAX_INT = 2**31 - 1
    MIN_INT = - 2**31

    ___ myAtoi  strs
        """ We need to handle four cases:

        1. discards all leading whitespaces
        2. sign of the number
        3. overflow
        4. invalid input
        """
        strs = strs.strip()
        __ n.. strs:
            r_ 0

        sign, i = 1, 0
        __ strs[i] __ '+':
            i += 1
        ____ strs[i] __ '-':
            i += 1
            sign = -1

        num = 0
        _____ i < l..(strs
            __ strs[i] < '0' or strs[i] > '9':
                break
            __ num > self.MAX_INT or (num * 10 + int(strs[i]) > self.MAX_INT
                r_ self.MAX_INT __ sign __ 1 else self.MIN_INT
            ____
                num = num * 10 + int(strs[i])
            i += 1

        r_ num * sign

"""
""
"  12a"
"  a12"
"  +12"
"  +-12"
"2147483648"
"-2147483648"
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/08_StringtoInteger.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-08-21 09:10:58
 */

c.. Solution {
public:
    int myAtoi(string str){
        /*
        We need to handle four cases:

        1. discards all leading whitespaces
        2. sign of the number
        3. overflow
        4. invalid input
        */
        const int l.. = str.size();

        int i=0;
        // Skip starting whitespace as many as possible.
        _____ (str[i]__' ' && i<l..) i++;

        int sign = 1;
        // Read plus or minus sign.
        __(str[i] __ '+') i++;
        else __(str[i] __ '-'){
            i++;
            sign = -1;
        }

        int num = 0;
        ___(; i<l..; i++){
            __(str[i] < '0' || str[i] > '9'){
                break;
            }
            __(num > INT_MAX/10 || (num __ INT_MAX / 10 && (str[i]-'0') > INT_MAX % 10)){
                r_ sign __ -1 ? INT_MIN : INT_MAX;
            }
            num = num * 10 + str[i] - '0';
        }

        r_ num * sign;
    }
};

/*
""
"  12a"
"  a12"
"  +12"
"  +-12"
"2147483648"
"-2147483648"
*/
